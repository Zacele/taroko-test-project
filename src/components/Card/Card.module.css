.contactCard {
  border: 1px solid black;
  padding: 16px;
  margin-bottom: 8px;
  width: 100%;
}

.highlighted {
  border-color: blue;
}

.header {
  display: flex;
  align-items: center;
}

.star {
  color: gold;
  margin-right: 8px;
}

.content {
  margin-top: 8px;
}

.actions {
  margin-top: 16px;
  text-align: right;
}

.actions button {
  margin-left: 8px;
}
