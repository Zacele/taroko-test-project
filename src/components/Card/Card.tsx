'use client';
import React from 'react';
import styles from './Card.module.css';

export interface CardProps {
  name: string;
  job: string;
  description: string;
  isHighlighted?: boolean;
  onDelete?: () => void;
}

const Card: React.FC<CardProps> = ({
  name,
  job,
  description,
  isHighlighted,
}) => {
  const onEdit = () => {
    console.log('Edit');
  };
  const onDelete = () => {
    console.log('Delete');
  };
  return (
    <div
      className={`${styles.contactCard} ${isHighlighted ? styles.highlighted : ''}`}
    >
      <div className={styles.header}>
        {isHighlighted && <span className={styles.star}>â˜…</span>}
        <h2>{name}</h2>
      </div>
      <div className={styles.content}>
        <p>Job: {job}</p>
        <p>Description: {description}</p>
      </div>
      <div className={styles.actions}>
        <button onClick={onEdit}>Edit</button>
        <button onClick={onDelete}>Delete</button>
      </div>
    </div>
  );
};

export default Card;
