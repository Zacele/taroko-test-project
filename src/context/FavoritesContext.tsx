'use client';
import React, {
  createContext,
  useContext,
  useState,
  ReactNode,
  useEffect,
} from 'react';
import Cookies from 'js-cookie';

type FavoritesContextType = {
  favorites: number[];
  toggleFavorite: (id: number) => void;
};

const FavoritesContext = createContext<FavoritesContextType | undefined>(
  undefined,
);

type FavoritesProviderProps = {
  children: ReactNode;
};

const FAVORITES_COOKIE_NAME = 'favorites';

export const FavoritesProvider: React.FC<FavoritesProviderProps> = ({
  children,
}) => {
  // Initialize state with empty array or from cookies
  const [favorites, setFavorites] = useState<number[]>(() => {
    const cookieValue = Cookies.get(FAVORITES_COOKIE_NAME);
    return cookieValue ? JSON.parse(cookieValue) : [];
  });

  useEffect(() => {
    // Update cookie whenever favorites change
    Cookies.set(FAVORITES_COOKIE_NAME, JSON.stringify(favorites), {
      expires: 7,
    });
  }, [favorites]);

  const toggleFavorite = (id: number) => {
    setFavorites((currentFavorites) => {
      const isFavorite = currentFavorites.includes(id);
      const newFavorites = isFavorite
        ? currentFavorites.filter((favId) => favId !== id)
        : [...currentFavorites, id];

      return newFavorites;
    });
  };

  return (
    <FavoritesContext.Provider value={{ favorites, toggleFavorite }}>
      {children}
    </FavoritesContext.Provider>
  );
};

export const useFavorites = (): FavoritesContextType => {
  const context = useContext(FavoritesContext);
  if (context === undefined) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};
