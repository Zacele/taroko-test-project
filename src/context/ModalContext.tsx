'use client';
import { ModalState } from '@/interface';
import React, { createContext, useContext, useState, ReactNode } from 'react';

const ModalContext = createContext<ModalState | undefined>(undefined);

export const useModal = () => {
  const context = useContext(ModalContext);
  if (context === undefined) {
    throw new Error('useModal must be used within a ModalProvider');
  }
  return context;
};

export const ModalProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [open, setOpen] = useState(false);
  const [isEditMode, setIsEditMode] = useState(false);
  const [selectedId, setSelectedId] = useState<number | null>(null);

  const closeModal = () => {
    setOpen(false);
    setIsEditMode(false);
    setSelectedId(null);
  };

  const openCreateModal = () => {
    setOpen(true);
  };

  const openEditModal = (id: number) => {
    setIsEditMode(true);
    setOpen(true);
    setSelectedId(id);
  };

  return (
    <ModalContext.Provider
      value={{
        isOpen: open,
        isEdit: isEditMode,
        openModal: openCreateModal,
        closeModal,
        openEditModal,
        selectedId,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};
