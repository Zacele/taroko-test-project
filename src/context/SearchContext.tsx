'use client';
import { ContactType } from '@/interface';
import React, { createContext, useContext, ReactNode, useState } from 'react';

interface SearchContextType {
  searchQuery: string;
  setSearchQuery: (query: string) => void;
  filterContacts: (contacts: ContactType[]) => ContactType[];
}

const SearchContext = createContext<SearchContextType | undefined>(undefined);

interface SearchProviderProps {
  children: ReactNode;
}

export const useSearch = () => {
  const context = useContext(SearchContext);
  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }
  return context;
};

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [searchQuery, setSearchQuery] = useState('');

  const filterContacts = (contacts: ContactType[]): ContactType[] => {
    return contacts.filter((contact) =>
      `${contact.first_name} ${contact.last_name}`
        .toLowerCase()
        .includes(searchQuery.toLowerCase()),
    );
  };

  return (
    <SearchContext.Provider
      value={{ searchQuery, setSearchQuery, filterContacts }}
    >
      {children}
    </SearchContext.Provider>
  );
};
