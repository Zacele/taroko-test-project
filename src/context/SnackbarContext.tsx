'use client';
import React, { createContext, useContext, useState, ReactNode } from 'react';
import Snackbar from '@mui/material/Snackbar';
import Alert, { AlertColor } from '@mui/material/Alert';

type SnackbarContextType = {
  showSnackbar: (message: string, severity: AlertColor) => void;
};

const SnackbarContext = createContext<SnackbarContextType | undefined>(
  undefined,
);

type Props = {
  children: ReactNode;
};

export const SnackbarProvider: React.FC<Props> = ({ children }) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState('');
  const [severity, setSeverity] = useState<AlertColor>('info');

  const showSnackbar = (message: string, severity: AlertColor = 'info') => {
    setMessage(message);
    setSeverity(severity);
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <SnackbarContext.Provider value={{ showSnackbar }}>
      {children}
      <Snackbar open={open} autoHideDuration={2000} onClose={handleClose}>
        <Alert onClose={handleClose} severity={severity} sx={{ width: '100%' }}>
          {message}
        </Alert>
      </Snackbar>
    </SnackbarContext.Provider>
  );
};

export const useSnackbar = () => {
  const context = useContext(SnackbarContext);
  if (context === undefined) {
    throw new Error('useSnackbar must be used within a SnackbarProvider');
  }
  return context;
};
