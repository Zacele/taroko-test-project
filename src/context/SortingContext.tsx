'use client';
import { SortingOrderType } from '@/interface';
import React, { createContext, useContext, useState, ReactNode } from 'react';

interface SortingContextType {
  order: SortingOrderType;
  setOrder: (order: SortingOrderType) => void;
}

const SortingContext = createContext<SortingContextType | undefined>(undefined);

interface SortingProviderProps {
  children: ReactNode;
}

export const SortingProvider: React.FC<SortingProviderProps> = ({
  children,
}) => {
  const [order, setOrder] = useState<SortingOrderType>('desc');

  return (
    <SortingContext.Provider value={{ order, setOrder }}>
      {children}
    </SortingContext.Provider>
  );
};

export const useSortingContext = () => {
  const context = useContext(SortingContext);
  if (context === undefined) {
    throw new Error('useSortingContext must be used within a SortingProvider');
  }
  return context;
};
