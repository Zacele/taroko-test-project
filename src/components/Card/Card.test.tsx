import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Card from './Card';
import * as ModalContext from '@/context/ModalContext';
import * as SnackbarContext from '@/context/SnackbarContext';
import * as FavoritesContext from '@/context/FavoritesContext';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

jest.mock('@/api', () => ({
  deleteContactById: jest.fn(),
}));

jest.mock('@/context/ModalContext', () => ({
  useModal: jest.fn(),
}));

jest.mock('@/context/SnackbarContext', () => ({
  useSnackbar: jest.fn(),
}));

jest.mock('@/context/FavoritesContext', () => ({
  useFavorites: jest.fn(),
}));

const renderWithProviders = (component: React.ReactNode) => {
  const queryClient = new QueryClient();
  return render(
    <QueryClientProvider client={queryClient}>{component}</QueryClientProvider>,
  );
};

describe('Card Component', () => {
  beforeEach(() => {
    (ModalContext.useModal as jest.Mock).mockReturnValue({
      openEditModal: jest.fn(),
    });

    (SnackbarContext.useSnackbar as jest.Mock).mockReturnValue({
      showSnackbar: jest.fn(),
    });

    (FavoritesContext.useFavorites as jest.Mock).mockReturnValue({
      toggleFavorite: jest.fn(),
    });
  });

  test('renders Card with provided props and matches snapshot', () => {
    const { asFragment } = renderWithProviders(
      <Card
        name="Yoda"
        job="Master Jedi"
        description="Stay in the hut."
        id={1}
        isHighlighted={false}
      />,
    );
    expect(screen.getByText('Yoda')).toBeInTheDocument();
    expect(screen.getByText('Job: Master Jedi')).toBeInTheDocument();
    expect(
      screen.getByText('Description: Stay in the hut.'),
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('calls openEditModal on edit button click and matches snapshot', () => {
    const { asFragment } = renderWithProviders(
      <Card
        name="John Doe"
        job="Developer"
        description="Writes code."
        id={1}
        isHighlighted={false}
      />,
    );
    fireEvent.click(screen.getByText('Edit'));
    expect(ModalContext.useModal().openEditModal).toHaveBeenCalledWith(1);
    expect(asFragment()).toMatchSnapshot();
  });

  test('calls deleteContactById on delete button click and matches snapshot', async () => {
    const { deleteContactById } = jest.requireMock('@/api');
    deleteContactById.mockImplementation(() => Promise.resolve()); // Simulate successful deletion

    const { asFragment } = renderWithProviders(
      <Card
        name="Obi-wan Kenobi"
        job="Frienemies with Anakin."
        description="In pain."
        id={2}
        isHighlighted={false}
      />,
    );

    fireEvent.click(screen.getByText('Delete'));

    await waitFor(() => {
      expect(deleteContactById).toHaveBeenCalledWith(2);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
