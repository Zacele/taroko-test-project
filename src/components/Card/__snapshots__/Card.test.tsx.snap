// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component calls deleteContactById on delete button click and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="contactCard"
  >
    <div
      class="header"
    >
      <span
        class="star"
      >
        ★
      </span>
      <h2>
        Obi-wan Kenobi
      </h2>
    </div>
    <div
      class="content"
    >
      <p>
        Job: Frienemies with Anakin.
      </p>
      <p>
        Description: In pain.
      </p>
    </div>
    <div
      class="actions"
    >
      <button
        class="button default"
      >
        Edit
      </button>
      <button
        class="button danger"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Card Component calls openEditModal on edit button click and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="contactCard"
  >
    <div
      class="header"
    >
      <span
        class="star"
      >
        ★
      </span>
      <h2>
        John Doe
      </h2>
    </div>
    <div
      class="content"
    >
      <p>
        Job: Developer
      </p>
      <p>
        Description: Writes code.
      </p>
    </div>
    <div
      class="actions"
    >
      <button
        class="button default"
      >
        Edit
      </button>
      <button
        class="button danger"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Card Component renders Card with provided props and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="contactCard"
  >
    <div
      class="header"
    >
      <span
        class="star"
      >
        ★
      </span>
      <h2>
        Yoda
      </h2>
    </div>
    <div
      class="content"
    >
      <p>
        Job: Master Jedi
      </p>
      <p>
        Description: Stay in the hut.
      </p>
    </div>
    <div
      class="actions"
    >
      <button
        class="button default"
      >
        Edit
      </button>
      <button
        class="button danger"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
